aten/src/ATen/DeviceGuard.h

```diff
@@ -53,84 +66,91 @@ struct DeviceGuard {
   /// moved-from `DeviceGuard` is modified such that its destruction has no
   /// effect (does not reset the device).
   DeviceGuard(DeviceGuard&& other) noexcept {
+    // Reuse move assignment implementation
     *this = std::move(other);
   }
 
   /// Move-assigns this `DeviceGuard` from another `DeviceGuard`. The
   /// moved-from `DeviceGuard` is modified such that its destruction has no
   /// effect (does not reset the device).
   DeviceGuard& operator=(DeviceGuard&& other) noexcept {
-    this->original_index_ = other.original_index_;
-    this->last_index_ = other.last_index_;
-    // Set other's original index to the unspecified/default state, so that it
-    // doesn't also reset the device in its constructor.
-    other.original_index_ = -1;
+    // We cannot use the default move assignment here.  Quoth the standard:
+    //
+    //    constexpr optional( optional&& other )
+    //
+    //    If other contains a value, then depending on whether *this contains a
+    //    value, the contained value is either direct-initialized or assigned from
+    //    *other (2) or std::move(*other) (3). Note that a moved-from optional
+    //    still contains a value.
+    //
+    // Swapping works fine though.
+    std::swap(this->impl_, other.impl_);
+    std::swap(this->original_device_, other.original_device_);
+    std::swap(this->current_device_, other.current_device_);
```

Comment:
If `other` continues to live on after the move assignment, this can cause weird behavior. In RAII classes, move assignment and construction should always make sure to set `other` to a clearly defined `uninitialized` state.